### PR TITLE
MSP430F5529 HAL

### PR DESCRIPTION
**Pull request by rplauche**
Generic MSP430F5529 HAL
* UART Module
* GPIO Module
* Base doxygen docs

Part of openkosmosorg/kubos-hal-msp430f5529#1

### CODE REVIEW

File: msp430f5529-hal/uart.h [hunk 1]
```diff
@@ -148,6 +176,29 @@ uint8_t hal_uart_setup(hal_uart_handle * handle);
 void hal_uart_set_baudrate(hal_uart_handle * handle);
 
 /**
+  * @brief Low level hardware setup of UART parity.
+  * @param handle Instance of initilaized hal_uart_handle containing hardware
+  *               registers and config values.
+  */
+void hal_uart_set_parity(hal_uart_handle * handle);
```
Comment (marshall) [inline msp430f5529-hal/uart.h]: IMO these 4 functions should be private / static, they are helper functions for `hal_uart_setup`, and I don't expect they'd be very useful to downstream developers

File: msp430f5529-hal/uart.h [hunk 2]
```diff
@@ -61,30 +101,97 @@ typedef struct
...
+/**
+  * @brief Externally implemented interrupt processing function
+  * @param handle UART handle
+  */
+extern void hal_uart_interrupt(hal_uart_handle * handle);
+
+/**
+  * @brief Checks for existence of interrupt flag
+  */
+#define HAL_UART_INT_FLAG(handle, flag) (flag == (flag & handle->reg->interruptFlags))
```
Comment (marshall) [inline msp430f5529-hal/uart.h]: funny, the STM32CubeF4 HAL has a similar macro that's a bit more general purpose called `CHECK_BIT`. We should make these a part of our APIs at some point, but it's not important right now

File: source/uart.c [hunk 3]
```diff
@@ -19,69 +19,54 @@
...
 #include "msp430f5529-hal/uart.h"
-
 #include <stddef.h>
+#include <stdbool.h>
 
-#define BC_RX_WAKE_THRESH 1
-
-static hal_uart_handle hal_uart_a0;
-static hal_uart_handle hal_uart_a1;
+hal_uart_handle hal_uart_dev[YOTTA_CFG_HARDWARE_UARTCOUNT];
```
Comment (marshall) [inline source/uart.c]: use `K_NUM_UARTS` here from `kubos-hal/uart.h`, which has some logic around this config already

File: source/uart.c [hunk 4]
```diff
@@ -19,69 +19,54 @@
...
+    hal_uart_handle * handle = NULL;
+
+    handle = &hal_uart_dev[device];
+    if (HAL_UART_A0 == device)
+    {
+        handle->config.device = device;
+        handle->select = &P3SEL;
+        handle->selectVal = BIT3 + BIT4;
+        handle->reg = (hal_uart_mem_reg *)__MSP430_BASEADDRESS_USCI_A0__;
+    } else if (HAL_UART_A1 == device)
```
Comment (marshall) [inline source/uart.c]: nit: I've noticed in some places you're putting `else if (..)` on a separate line from the closing bracket of the previous `if`, and here you're putting it on the same line as the bracket.

In the near future I want to decide on source code formatting standards, but for now let's at least be internally consistent :)

File: source/uart.c [hunk 5]
```diff
@@ -120,79 +106,112 @@ void hal_uart_set_speed(hal_uart_handle * handle)
...
-    count = handle->rxBufferIndex;
-    handle->rxBufferIndex = 0;
 
-    handle->reg->interruptFlags |= UCRXIE;
+void hal_uart_write(hal_uart_handle * handle, uint8_t c)
+{
+    hal_uart_write_raw(handle, c);
+    // Wait until bit has been clocked out...
+    while (!(UCTXIFG == (UCTXIFG & (handle->reg->interruptFlags)))
+        && (UCBUSY == ((handle->reg->status) & UCBUSY)));
```
Comment (marshall) [inline source/uart.c]: nit: you made 2 macros `HAL_UART_STAT` and `HAL_UART_INT_FLAG` to make this code less ugly, and didn't use them here :)

File: source/uart.c (hunk 3 above)
Comment (rplauche) [inline source/uart.c]: I was attempting to keep this module away from a dependency on the kubos-hal module.

File: msp430f5529-hal/uart.h (hunk 1 above)
Comment (rplauche) [inline msp430f5529-hal/uart.h]: Good call, fixed.

File: source/uart.c (hunk 4 above)
Comment (rplauche) [inline source/uart.c]: Fixed.

File: source/uart.c (hunk 5 above)
Comment (rplauche) [inline source/uart.c]: Fixed.

**Pull request merged**